src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/NuGetSdkResolver.cs

```diff
@@ -130,15 +131,16 @@ public static SdkResult GetSdkResult(SdkReference sdk, object nuGetVersion, SdkR
                         DefaultCredentialServiceUtility.SetupDefaultCredentialService(nugetSDKLogger, nonInteractive: true);
 
                         // Asynchronously run the restore without a commit which find the package on configured feeds, download, and unzip it without generating any other files
-                        var results = RestoreRunnerEx.RunWithoutCommit(
-                                context.ProjectFilePath,
-                                sdk.Name,
-                                parsedSdkVersion.ToFullString(),
-                                settings,
-                                nugetSDKLogger)
-                            .ConfigureAwait(continueOnCapturedContext: false)
-                            .GetAwaiter()
-                            .GetResult();
+                        // This must be run in its own task because legacy project system evaluates projects on the UI thread which can cause RunWithoutCommit() to deadlock
+                        // https://developercommunity.visualstudio.com/content/problem/311379/solution-load-never-completes-when-project-contain.html
+                        var restoreTask = Task.Run(() => RestoreRunnerEx.RunWithoutCommit(
+                            context.ProjectFilePath,
+                            sdk.Name,
+                            parsedSdkVersion.ToFullString(),
+                            settings,
+                            nugetSDKLogger));
+
+                        var results = restoreTask.Result;
```

Comment:
use `await restroeTask` instead of `Result` which is more prone to deadlock

Reply:
This method implements an MSBuild interface that is not marked `async`, so `await` isn't going to work.

Reply:
ahh then can you use `JoinableTaskFactory` there?

Reply:
I wasn't sure if taking a dependency on JTF in just the SDK resolver would fix the issue.  `RunWithoutCommit` is async and it does some fancy stuff with tasks: https://github.com/NuGet/NuGet.Client/blob/f68e847f4e33164769050889dec6b426c1e39b97/src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs#L128

And then the API in MSBuild which calls all of this is not async.  So this seemed like the easiest way to get it fixed.

Reply:
yeah I can understand that.